service/frontend/nexus_incoming_service_client.go

```diff
@@ -0,0 +1,373 @@
+	if entry.Version > 1 {
+		lastModifiedTime = timestamppb.New(hlc.UTC(entry.Service.Clock))
+	}
+
+	return &nexus.IncomingService{
+		Version:          entry.Version,
+		Id:               entry.Id,
+		Spec:             entry.Service.Spec,
+		CreatedTime:      entry.Service.CreatedTime,
+		LastModifiedTime: lastModifiedTime,
+		UrlPrefix:        "/" + cnexus.Routes().DispatchNexusTaskByService.Path(entry.Id),
+	}
+}
+
+func (c *NexusIncomingServiceClient) getServiceNameIssues(name string) rpc.RequestIssues {
+	var issues rpc.RequestIssues
+
+	if name == "" {
+		issues.Append("incoming service name not set")
+		return issues
+	}
+
+	maxNameLength := c.config.maxNameLength()
+	if len(name) > maxNameLength {
+		issues.Appendf("incoming service name exceeds length limit of %d", maxNameLength)
+	}
+
+	if !ServiceNameRegex.MatchString(name) {
+		issues.Appendf("incoming service name must match the regex: %q", ServiceNameRegex.String())
+	}
+
+	return issues
+}
+
+func (c *NexusIncomingServiceClient) validateUpsertSpec(spec *nexus.IncomingServiceSpec) error {
+	issues := c.getServiceNameIssues(spec.GetName())
+
+	if spec.Namespace == "" {
+		issues.Append("incoming service namespace not set")
+	} else if _, nsErr := c.namespaceRegistry.GetNamespace(namespace.Name(spec.Namespace)); nsErr != nil {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("could not verify namespace referenced by incoming service exists: %v", nsErr.Error()))
+	}
+
+	if spec.TaskQueue == "" {
+		issues.Append("incoming service task queue not set")
+	} else {
+		maxTaskQueueLength := c.config.maxTaskQueueLength()
+		if len(spec.TaskQueue) > maxTaskQueueLength {
+			issues.Appendf("incoming service task queue exceeds length limit of %d", maxTaskQueueLength)
+		}
+		if strings.HasPrefix(spec.TaskQueue, reservedTaskQueuePrefix) {
+			issues.Appendf("incoming service task queue begins with reserved prefix: %q", reservedTaskQueuePrefix)
+		}
+	}
+
+	maxSize := c.config.maxSize()
+	if spec.Size() > maxSize {
+		issues.Appendf("incoming service size exceeds limit of %d", maxSize)
+	}
+
+	return issues.GetError()
+}
+
+func getServiceIdIssues(ID string) rpc.RequestIssues {
+	var issues rpc.RequestIssues
+	if ID == "" {
+		issues.Append("incoming service Id not set")
+	} else if _, err := uuid.Parse(ID); err != nil {
+		issues.Appendf("malformed incoming service Id: %q", err.Error())
+	}
+	return issues
+}
+
+func validateDeleteRequest(request *operatorservice.DeleteNexusIncomingServiceRequest) error {
+	issues := getServiceIdIssues(request.GetId())
+
+	if request.GetVersion() <= 0 {
+		issues.Append("incoming service version is non-positive")
+	}
+
+	return issues.GetError()
+}
+
+func validateGetRequest(request *operatorservice.GetNexusIncomingServiceRequest) error {
+	issues := getServiceIdIssues(request.GetId())
+	return issues.GetError()
+}
+
+func (c *NexusIncomingServiceClient) validatePageSize(pageSize int32) error {
+	if pageSize < 0 {
```

Comment:
What about 0?

Reply:
We treat 0 as unset, so it will always be updated to the default and will not go through this validation.

Reply:
Wonder if it's worth a comment here or moving setting the default here to avoid making assumptions.